File: sdk/typescript/tests/media_provider.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { MediaRouter, MediaProviderError } from '../src/ai/MediaProvider.js';
import type { MediaProvider, MediaResponse } from '../src/ai/MediaProvider.js';
import { OpenRouterMediaProvider } from '../src/ai/OpenRouterMediaProvider.js';

// ── MediaRouter tests ────────────────────────────────────────────────

function makeStubProvider(
  name: string,
  modalities: string[] = ['image', 'audio', 'video']
): MediaProvider {
  const empty: MediaResponse = {
    text: '',
    images: [],
    audio: null,
    files: [],
    videos: [],
    rawResponse: null,
  };
  return {
    name,
    supportedModalities: modalities,
    generateImage: vi.fn().mockResolvedValue(empty),
    generateAudio: vi.fn().mockResolvedValue(empty),
    generateVideo: vi.fn().mockResolvedValue(empty),
  };
}

describe('MediaRouter', () => {
  it('resolves provider by prefix', () => {
    const router = new MediaRouter();
    const prov = makeStubProvider('openrouter');
    router.register('openrouter/', prov);
    expect(router.resolve('openrouter/google/veo-3', 'video')).toBe(prov);
  });

  it('longest prefix wins', () => {
    const router = new MediaRouter();
    const generic = makeStubProvider('generic');
    const specific = makeStubProvider('specific');
    router.register('openrouter/', generic);
    router.register('openrouter/google/', specific);

    expect(router.resolve('openrouter/google/veo-3', 'video')).toBe(specific);
    expect(router.resolve('openrouter/openai/dall-e', 'image')).toBe(generic);
  });

  it('throws MediaProviderError when no provider matches', () => {
    const router = new MediaRouter();
    expect(() => router.resolve('unknown/model', 'video')).toThrow(MediaProviderError);
    expect(() => router.resolve('unknown/model', 'video')).toThrow(
      "No provider for model 'unknown/model' with 'video' capability"
    );
  });

  it('checks capability filter', () => {
    const router = new MediaRouter();
    const imageOnly = makeStubProvider('img', ['image']);
    router.register('img/', imageOnly);

    expect(router.resolve('img/model', 'image')).toBe(imageOnly);
    expect(() => router.resolve('img/model', 'video')).toThrow(MediaProviderError);
  });
});

// ── MediaProviderError tests ────────────────────────────────────────

describe('MediaProviderError', () => {
  it('is an instance of Error', () => {
    const err = new MediaProviderError('test');
    expect(err).toBeInstanceOf(Error);
    expect(err).toBeInstanceOf(MediaProviderError);
    expect(err.name).toBe('MediaProviderError');
  });

  it('carries structured context', () => {
    const err = new MediaProviderError('fail', {
      provider: 'openrouter',
      model: 'google/veo-3',
      endpoint: '/api/v1/videos',
    });
    expect(err.provider).toBe('openrouter');
    expect(err.model).toBe('google/veo-3');
    expect(err.endpoint).toBe('/api/v1/videos');
  });
});

// ── OpenRouterMediaProvider tests ────────────────────────────────────

describe('OpenRouterMediaProvider', () => {
  const originalFetch = globalThis.fetch;
  let mockFetch: ReturnType<typeof vi.fn>;

  beforeEach(() => {
    mockFetch = vi.fn();
    globalThis.fetch = mockFetch;
  });

  afterEach(() => {
    globalThis.fetch = originalFetch;
  });

  it('throws MediaProviderError without API key', () => {
    const orig = process.env.OPENROUTER_API_KEY;
    delete process.env.OPENROUTER_API_KEY;
    expect(() => new OpenRouterMediaProvider()).toThrow(MediaProviderError);
    expect(() => new OpenRouterMediaProvider()).toThrow('API key required');
    if (orig) process.env.OPENROUTER_API_KEY = orig;
  });

  it('accepts apiKey in constructor', () => {
    const p = new OpenRouterMediaProvider({ apiKey: 'test-key' });
    expect(p.name).toBe('openrouter');
    expect(p.supportedModalities).toContain('video');
  });

  it('includes OpenRouter attribution headers in fetch calls', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({ choices: [] }),
    });

    const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });
    await provider.generateImage({ prompt: 'test' });

    const init = mockFetch.mock.calls[0][1];
    expect(init.headers).toEqual(
      expect.objectContaining({
        Authorization: 'Bearer test-key',
        'HTTP-Referer': 'https://agentfield.ai',
        'X-OpenRouter-Title': 'AgentField AI',
        'X-Title': 'AgentField AI',
      })
    );
  });

  it('toJSON excludes API key (CR-03)', () => {
    const p = new OpenRouterMediaProvider({ apiKey: 'secret-key-123' });
    const json = JSON.parse(JSON.stringify(p));
    expect(json).not.toHaveProperty('apiKey');
    expect(Object.values(json).join(' ')).not.toContain('secret-key-123');
    expect(json.name).toBe('openrouter');
    expect(json.baseUrl).toBeDefined();
  });

  it('API key not on instance properties (CR-03)', () => {
    const p = new OpenRouterMediaProvider({ apiKey: 'secret-key-123' });
    const keys = Object.keys(p);
    expect(keys).not.toContain('apiKey');
    // Ensure key isn't accessible via any enumerable property
    for (const k of keys) {
      expect((p as any)[k]).not.toBe('secret-key-123');
    }
  });

  it('strips openrouter/ prefix from model', async () => {
    const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

    // Mock image generation
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        choices: [
          {
            message: {
              content: [
                { type: 'text', text: 'Generated' },
                { type: 'image_url', image_url: { url: 'https://example.com/img.png' } },
              ],
            },
          },
        ],
      }),
    });

    await provider.generateImage({ prompt: 'a cat', model: 'openrouter/openai/gpt-image-1' });

    // Verify the model sent in the body has prefix stripped
    const callBody = JSON.parse(mockFetch.mock.calls[0][1].body);
    expect(callBody.model).toBe('openai/gpt-image-1');
  });

  it('passes AbortSignal.timeout to fetch calls (CR-01)', async () => {
    const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => ({
        choices: [{ message: { content: 'test' } }],
      }),
    });

    await provider.generateImage({ prompt: 'test' });

    // Verify signal is present on the fetch call
    const fetchOptions = mockFetch.mock.calls[0][1];
    expect(fetchOptions.signal).toBeDefined();
  });

  describe('generateVideo', () => {
    it('submits job, polls, downloads', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });
      const videoBytes = Buffer.from('fake-video-data');

      // Submit
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ id: 'job-123' }),
      });
      // Poll -> completed
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          id: 'job-123',
          status: 'completed',
          unsigned_url: 'https://example.com/video.mp4',
          cost_usd: 0.05,
        }),
      });
      // Download
      mockFetch.mockResolvedValueOnce({
        ok: true,
        arrayBuffer: async () => videoBytes.buffer.slice(
          videoBytes.byteOffset,
          videoBytes.byteOffset + videoBytes.byteLength
        ),
      });

      const resp = await provider.generateVideo({
        prompt: 'a sunset',
        model: 'openrouter/google/veo-3',
        pollInterval: 1, // 1ms for test speed
      });

      expect(resp.videos).toHaveLength(1);
      expect(resp.videos[0].url).toBe('https://example.com/video.mp4');
      expect(resp.videos[0].data).toBe(videoBytes.toString('base64'));
      expect(resp.videos[0].costUsd).toBe(0.05);

      // Verify calls: submit, poll, download
      expect(mockFetch).toHaveBeenCalledTimes(3);

      // Verify download fetch has redirect: 'error' (CR-02)
      const downloadOptions = mockFetch.mock.calls[2][1];
      expect(downloadOptions.redirect).toBe('error');
    });

    it('throws on submit failure with context (WR-05)', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 402,
        text: async () => 'Insufficient credits',
      });

      await expect(
        provider.generateVideo({ prompt: 'test', pollInterval: 1 })
      ).rejects.toThrow('Video submit failed');
    });

    it('submit failure is MediaProviderError', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 402,
        text: async () => 'Insufficient credits',
      });

      await expect(
        provider.generateVideo({ prompt: 'test', pollInterval: 1 })
      ).rejects.toThrow(MediaProviderError);
    });

    it('throws on generation failure status', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ id: 'job-fail' }),
      });
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ id: 'job-fail', status: 'failed', error: 'content policy' }),
      });

      await expect(
        provider.generateVideo({ prompt: 'test', pollInterval: 1 })
      ).rejects.toThrow('Video generation failed');
    });

    it('throws on timeout', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ id: 'job-slow' }),
      });
      // Always return pending
      mockFetch.mockResolvedValue({
        ok: true,
        json: async () => ({ id: 'job-slow', status: 'pending' }),
      });

      await expect(
        provider.generateVideo({ prompt: 'test', pollInterval: 1, timeout: 10 })
      ).rejects.toThrow('timed out');
    });

    it('rejects non-https video download URL (CR-02)', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      // Submit
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ id: 'job-http' }),
      });
      // Poll -> completed with http URL
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          id: 'job-http',
          status: 'completed',
          unsigned_url: 'http://example.com/video.mp4',
        }),
      });

      await expect(
        provider.generateVideo({ prompt: 'test', pollInterval: 1 })
      ).rejects.toThrow('non-HTTPS');
    });

    it('rejects localhost video download URL (CR-02)', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ id: 'job-local' }),
      });
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          id: 'job-local',
          status: 'completed',
          unsigned_url: 'https://localhost/video.mp4',
        }),
      });

      await expect(
        provider.generateVideo({ prompt: 'test', pollInterval: 1 })
      ).rejects.toThrow('localhost');
    });

    it('rejects private IP video download URL (CR-02)', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({ id: 'job-priv' }),
      });
      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          id: 'job-priv',
          status: 'completed',
          unsigned_url: 'https://10.0.0.1/video.mp4',
        }),
      });

      await expect(
        provider.generateVideo({ prompt: 'test', pollInterval: 1 })
      ).rejects.toThrow('private IP');
    });
  });

  describe('generateImage', () => {
    it('extracts images from response', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          choices: [
            {
              message: {
                content: [
                  { type: 'text', text: 'Here is your image' },
                  { type: 'image_url', image_url: { url: 'https://cdn.example.com/img.png' } },
                ],
              },
            },
          ],
        }),
      });

      const resp = await provider.generateImage({ prompt: 'a cat' });
      expect(resp.text).toBe('Here is your image');
      expect(resp.images).toHaveLength(1);
      expect(resp.images[0].url).toBe('https://cdn.example.com/img.png');
    });

    it('handles data URL images with b64', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });

      mockFetch.mockResolvedValueOnce({
        ok: true,
        json: async () => ({
          choices: [
            {
              message: {
                content: [
                  {
                    type: 'image_url',
                    image_url: { url: 'data:image/png;base64,iVBOR' },
                  },
                ],
              },
            },
          ],
        }),
      });

      const resp = await provider.generateImage({ prompt: 'test' });
      expect(resp.images[0].b64Json).toBe('iVBOR');
    });
  });

  describe('generateAudio', () => {
    it('parses SSE stream and collects audio chunks', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });
      // Force chat-completions routing — without this, the provider would try
      // GET /models/{id}/endpoints to discover routing and fall through to
      // /audio/speech because mocks return undefined.
      provider.seedModelMeta('openai/gpt-audio-mini', ['text', 'audio'], ['text']);

      const sseLines = [
        'data: {"choices":[{"delta":{"content":"Hello"}}]}\n\n',
        'data: {"choices":[{"delta":{"audio":{"data":"AAAA"}}}]}\n\n',
        'data: {"choices":[{"delta":{"audio":{"data":"BBBB"}}}]}\n\n',
        'data: [DONE]\n\n',
      ];
      const encoder = new TextEncoder();
      let callIndex = 0;

      const mockReader = {
        read: vi.fn().mockImplementation(async () => {
          if (callIndex < sseLines.length) {
            const chunk = encoder.encode(sseLines[callIndex]);
            callIndex++;
            return { done: false, value: chunk };
          }
          return { done: true, value: undefined };
        }),
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        body: { getReader: () => mockReader },
      });

      // mp3 keeps chunks as raw base64 (no wav wrap).
      const resp = await provider.generateAudio({
        text: 'say hello', model: 'openai/gpt-audio-mini', format: 'mp3',
      });
      expect(resp.text).toBe('Hello');
      expect(resp.audio).not.toBeNull();
      expect(resp.audio!.data).toBe('AAAABBBB');
      expect(resp.audio!.format).toBe('mp3');
    });

    it('processes remaining buffer after stream ends (WR-02)', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });
      provider.seedModelMeta('openai/gpt-audio-mini', ['text', 'audio'], ['text']);

      // Send data without trailing newline so it stays in buffer
      const encoder = new TextEncoder();
      let callIndex = 0;
      const chunks = [
        'data: {"choices":[{"delta":{"content":"A"}}]}\n\n',
        'data: {"choices":[{"delta":{"content":"B"}}]}', // no trailing newline
      ];

      const mockReader = {
        read: vi.fn().mockImplementation(async () => {
          if (callIndex < chunks.length) {
            const chunk = encoder.encode(chunks[callIndex]);
            callIndex++;
            return { done: false, value: chunk };
          }
          return { done: true, value: undefined };
        }),
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        body: { getReader: () => mockReader },
      });

      const resp = await provider.generateAudio({
        text: 'test', model: 'openai/gpt-audio-mini', format: 'mp3',
      });
      expect(resp.text).toBe('AB');
    });

    it('throws on failure with context (WR-05)', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });
      provider.seedModelMeta('openai/gpt-audio-mini', ['text', 'audio'], ['text']);

      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 500,
        text: async () => 'Internal error',
      });

      await expect(
        provider.generateAudio({
          text: 'test', model: 'openai/gpt-audio-mini', format: 'mp3',
        })
      ).rejects.toThrow(MediaProviderError);
      // Reset mock for second assertion
      mockFetch.mockResolvedValueOnce({
        ok: false,
        status: 500,
        text: async () => 'Internal error',
      });
      await expect(
        provider.generateAudio({
          text: 'test', model: 'openai/gpt-audio-mini', format: 'mp3',
        })
      ).rejects.toThrow('Audio generation failed');
    });

    it('throws after too many consecutive parse errors (WR-04)', async () => {
      const provider = new OpenRouterMediaProvider({ apiKey: 'test-key' });
      provider.seedModelMeta('openai/gpt-audio-mini', ['text', 'audio'], ['text']);

      // Create 51+ malformed SSE lines in a single chunk
      const malformedLines = Array.from(
        { length: 55 },
        (_, i) => `data: NOT-VALID-JSON-${i}`
      ).join('\n') + '\n';

      const encoder = new TextEncoder();
      let sent = false;

      const mockReader = {
        read: vi.fn().mockImplementation(async () => {
          if (!sent) {
            sent = true;
            return { done: false, value: encoder.encode(malformedLines) };
          }
          return { done: true, value: undefined };
        }),
      };

      mockFetch.mockResolvedValueOnce({
        ok: true,
        body: { getReader: () => mockReader },
      });

      await expect(
        provider.generateAudio({
          text: 'test', model: 'openai/gpt-audio-mini', format: 'mp3',
        })
      ).rejects.toThrow('consecutive SSE parse errors');
    });
  });
});

File: sdk/go/ai/media_provider_test.go
package ai

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// mockProvider implements MediaProvider for testing.
type mockProvider struct {
	name       string
	modalities []string
}

func (m *mockProvider) Name() string                { return m.name }
func (m *mockProvider) SupportedModalities() []string { return m.modalities }
func (m *mockProvider) GenerateImage(_ context.Context, _ ImageRequest) (*MediaResponse, error) {
	return &MediaResponse{Text: m.name + ":image"}, nil
}
func (m *mockProvider) GenerateAudio(_ context.Context, _ AudioRequest) (*MediaResponse, error) {
	return &MediaResponse{Text: m.name + ":audio"}, nil
}
func (m *mockProvider) GenerateVideo(_ context.Context, _ VideoRequest) (*MediaResponse, error) {
	return &MediaResponse{Text: m.name + ":video"}, nil
}

func TestMediaRouterResolve(t *testing.T) {
	router := NewMediaRouter()

	or := &mockProvider{name: "openrouter", modalities: []string{"image", "audio", "video"}}
	other := &mockProvider{name: "other", modalities: []string{"image"}}

	router.Register("openrouter/", or)
	router.Register("other/", other)

	tests := []struct {
		model      string
		capability string
		wantName   string
		wantErr    bool
	}{
		{"openrouter/kling", "video", "openrouter", false},
		{"openrouter/gpt-image-1", "image", "openrouter", false},
		{"other/some-model", "image", "other", false},
		{"other/some-model", "video", "", true},  // other doesn't support video
		{"unknown/model", "image", "", true},       // no matching prefix
	}

	for _, tt := range tests {
		t.Run(fmt.Sprintf("%s_%s", tt.model, tt.capability), func(t *testing.T) {
			p, err := router.Resolve(tt.model, tt.capability)
			if tt.wantErr {
				assert.Error(t, err)
				assert.Nil(t, p)
			} else {
				require.NoError(t, err)
				assert.Equal(t, tt.wantName, p.Name())
			}
		})
	}
}

func TestMediaRouterLongestPrefixMatch(t *testing.T) {
	router := NewMediaRouter()

	general := &mockProvider{name: "general", modalities: []string{"image"}}
	specific := &mockProvider{name: "specific", modalities: []string{"image"}}

	router.Register("openrouter/", general)
	router.Register("openrouter/dall-e", specific)

	p, err := router.Resolve("openrouter/dall-e-3", "image")
	require.NoError(t, err)
	assert.Equal(t, "specific", p.Name(), "longer prefix should match first")

	p, err = router.Resolve("openrouter/kling-v2", "image")
	require.NoError(t, err)
	assert.Equal(t, "general", p.Name(), "shorter prefix should match as fallback")
}

func TestMediaRouterEmpty(t *testing.T) {
	router := NewMediaRouter()
	_, err := router.Resolve("any/model", "image")
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "no provider")
}

func TestStripPrefix(t *testing.T) {
	tests := []struct {
		input string
		want  string
	}{
		{"openrouter/kling-video/v2", "kling-video/v2"},
		{"openrouter/gpt-image-1", "gpt-image-1"},
		{"plain-model", "plain-model"},
		{"openrouter/", ""},
	}
	for _, tt := range tests {
		assert.Equal(t, tt.want, stripPrefix(tt.input))
	}
}

func TestOpenRouterMediaProviderName(t *testing.T) {
	p, err := NewOpenRouterMediaProvider("test-key")
	require.NoError(t, err)
	assert.Equal(t, "openrouter", p.Name())
	assert.Equal(t, []string{"image", "audio", "video"}, p.SupportedModalities())
}

func TestOpenRouterMediaProviderDefaultKey(t *testing.T) {
	t.Setenv("OPENROUTER_API_KEY", "env-key")
	p, err := NewOpenRouterMediaProvider("")
	require.NoError(t, err)
	assert.Equal(t, "env-key", p.APIKey)
}

func TestOpenRouterMediaProviderEmptyKey(t *testing.T) {
	t.Setenv("OPENROUTER_API_KEY", "")
	_, err := NewOpenRouterMediaProvider("")
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "API key required")
}

func TestOpenRouterGenerateImage(t *testing.T) {
	// Mock server returning a chat completion with image content
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/chat/completions", r.URL.Path)
		assert.Equal(t, "Bearer test-key", r.Header.Get("Authorization"))

		var payload map[string]any
		require.NoError(t, json.NewDecoder(r.Body).Decode(&payload))
		assert.Equal(t, "gpt-image-1", payload["model"])

		resp := map[string]any{
			"choices": []map[string]any{
				{
					"message": map[string]any{
						"content": []map[string]any{
							{"type": "text", "text": "Here is your image"},
							{"type": "image_url", "b64_json": "aW1hZ2VkYXRh"},
						},
					},
				},
			},
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(resp)
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{
		APIKey:  "test-key",
		BaseURL: srv.URL,
		Client:  srv.Client(),
	}

	resp, err := p.GenerateImage(context.Background(), ImageRequest{
		Prompt: "a cat",
		Model:  "openrouter/gpt-image-1",
	})
	require.NoError(t, err)
	assert.Equal(t, "Here is your image", resp.Text)
	require.Len(t, resp.Images, 1)
	assert.Equal(t, "aW1hZ2VkYXRh", resp.Images[0].B64JSON)
}

func TestOpenRouterGenerateImageError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		w.Write([]byte(`{"error": "bad request"}`))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{
		APIKey:  "test-key",
		BaseURL: srv.URL,
		Client:  srv.Client(),
	}

	_, err := p.GenerateImage(context.Background(), ImageRequest{Prompt: "test"})
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "400")
}

func TestOpenRouterGenerateAudio(t *testing.T) {
	// Mock SSE streaming server
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		assert.Equal(t, "/chat/completions", r.URL.Path)

		var payload map[string]any
		require.NoError(t, json.NewDecoder(r.Body).Decode(&payload))
		assert.Equal(t, true, payload["stream"])

		w.Header().Set("Content-Type", "text/event-stream")
		flusher, _ := w.(http.Flusher)

		// Send text chunk
		fmt.Fprintf(w, "data: %s\n\n", `{"choices":[{"delta":{"content":"Hello"}}]}`)
		flusher.Flush()

		// Send audio chunk (base64 of "audio")
		fmt.Fprintf(w, "data: %s\n\n", `{"choices":[{"delta":{"audio":{"data":"YXVkaW8="}}}]}`)
		flusher.Flush()

		fmt.Fprintf(w, "data: [DONE]\n\n")
		flusher.Flush()
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{
		APIKey:  "test-key",
		BaseURL: srv.URL,
		Client:  srv.Client(),
	}
	p.SeedModelMeta("openai/gpt-audio-mini", []string{"text", "audio"}, []string{"text"})

	resp, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text:   "Say hello",
		Model:  "openai/gpt-audio-mini",
		Voice:  "nova",
		Format: "mp3",
	})
	require.NoError(t, err)
	assert.Equal(t, "Hello", resp.Text)
	require.NotNil(t, resp.Audio)
	assert.Equal(t, "mp3", resp.Audio.Format)
	assert.NotEmpty(t, resp.Audio.Data)
}

func TestOpenRouterGenerateVideoSubmitError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusForbidden)
		w.Write([]byte(`{"error":"forbidden"}`))
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{
		APIKey:  "test-key",
		BaseURL: srv.URL,
		Client:  srv.Client(),
	}

	_, err := p.GenerateVideo(context.Background(), VideoRequest{
		Prompt: "test",
		Model:  "openrouter/kling",
	})
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "403")
}

func TestOpenRouterGenerateVideoFullLifecycle(t *testing.T) {
	callCount := 0
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch {
		case r.Method == http.MethodPost && r.URL.Path == "/videos":
			json.NewEncoder(w).Encode(map[string]string{"id": "job-123"})
		case r.Method == http.MethodGet && r.URL.Path == "/videos/job-123":
			callCount++
			if callCount == 1 {
				json.NewEncoder(w).Encode(map[string]any{
					"id": "job-123", "status": "processing",
				})
			} else {
				json.NewEncoder(w).Encode(map[string]any{
					"id":           "job-123",
					"status":       "completed",
					"unsigned_url": "https://example.com/video.mp4",
					"duration":     5.0,
					"cost_usd":     0.05,
				})
			}
		default:
			w.WriteHeader(http.StatusNotFound)
		}
	}))
	defer srv.Close()

	p := &OpenRouterMediaProvider{
		APIKey:  "test-key",
		BaseURL: srv.URL,
		Client:  srv.Client(),
	}

	resp, err := p.GenerateVideo(context.Background(), VideoRequest{
		Prompt:       "test video",
		Model:        "openrouter/kling",
		PollInterval: 10 * time.Millisecond,
		Timeout:      5 * time.Second,
	})
	require.NoError(t, err)
	require.Len(t, resp.Videos, 1)
	assert.Equal(t, "https://example.com/video.mp4", resp.Videos[0].URL)
	assert.Equal(t, "generated_video.mp4", resp.Videos[0].Filename)
	assert.Equal(t, 5.0, resp.Videos[0].Duration)
	assert.Equal(t, 0.05, resp.Videos[0].CostUSD)
	assert.Equal(t, "video/mp4", resp.Videos[0].MimeType)
}

func TestOpenRouterGenerateVideoEmptyPrompt(t *testing.T) {
	p := &OpenRouterMediaProvider{
		APIKey:  "test-key",
		BaseURL: "http://localhost",
		Client:  &http.Client{},
	}
	_, err := p.GenerateVideo(context.Background(), VideoRequest{
		Prompt: "",
		Model:  "openrouter/kling",
	})
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "prompt must not be empty")
}

func TestOpenRouterGenerateAudioEmptyText(t *testing.T) {
	p := &OpenRouterMediaProvider{
		APIKey:  "test-key",
		BaseURL: "http://localhost",
		Client:  &http.Client{},
	}
	_, err := p.GenerateAudio(context.Background(), AudioRequest{
		Text: "  ",
	})
	assert.Error(t, err)
	assert.Contains(t, err.Error(), "text input must not be empty")
}
